### ISSUE TITLE
The astype methods on datetime64 Index objects do not work as expected

### ISSUE THREAD

**Issue by chaburkland** [InvestmentSystems/static-frame#195]
When attempting to cast an `IndexYearMonth`'s dtype to str, it did not succeed, but it also did not raise:
```python
>>> idx = sf.IndexYearMonth.from_date_range(date(2019,1,1), date(2019, 3,1))
>>> idx
'''
<IndexYearMonth>
2019-01
2019-02
2019-03
<datetime64[M]>
'''
>>> idx.astype(str)
'''
<IndexYearMonth>
2019-01
2019-02
2019-03
<datetime64[M]>
'''
```

After investigating a bit more, it seemed that this cast was not meant to succeed, as idx is a `IndexYearMonth` object, and so it's dtype is fixed to `np.datetime64[M]`. If that is indeed the case, then it seems that calling `.astype` on Index objects with a locked dtype should raise an exception letting the user know this behavior is not allowed on certain Index subclasses.

From what I discovered, the only way to change the dtype of my IndexYearMonth index, was to cast the underlying numpy array, and build a new Index.
```python
>>> sf.Index(idx.values.astype(str))
'''
<Index>
2019-01
2019-02
2019-03
2019-04
<<U25>
'''
```
This discovery wasn't immediate, as I went through about 4 other options before I realized I needed to call `.values` directly:
```python
>>> sf.Index(idx, dtype=str) # Raises RuntimeError
>>> sf.Index(idx.astype(str)) # Builds an Index with dtype `datetime64[M]`
>>> sf.Index(idx.astype(str), dtype=str) # Raises RuntimeError
>>> sf.Index(idx.values, dtype=str) # Raises RuntimeError
```

 Given this, I think it at least warrants a discussion about extending the logic on `sf.Index.astype` to factor in the new index based on the current Index subclass and the given dtype.

For example:
`sf.Index.astype(str) -> sf.Index`
`sf.Index.astype('datetime64[M]') -> sf.IndexYearMonth`
`sf.IndexYearMonth.astype('datetime64[D]') -> sf.IndexDate`
`sf.IndexYearMonth.astype(str) -> sf.Index`
`sf.IndexYearMonth.astype(object) -> sf.Index`
etc...

A new interface is also a possibility, although I wasn't able to come up with a suggestion I really liked. To further the discussion though, here were the two I came up with:

1) `idx.index_astype(dtype=str)`
2) `idx.as_index_type(sf.Index, dtype=str)`

**Comment by flexatone** [comment]
I think I prefer your first suggestion: having the `astype()` methods be specialized on datetime64 Indices to actually return a different Index type. Let me explore that implementation. We do not (yet) have Index objects for every datetime64 type; in those cases it should probably just fail.

**Comment by chaburkland** [comment]
I like my first suggestion the most too! In my initial draft of this issue, I was leaning towards adding a new interface, but as I explored those possibilities, I found myself not liking any of them and instead wanting `astype()` itself to return the appropriate Index type.

At the least, (assuming nothing with `astype()` is changed), do you think it's reasonable to have Index date-like subclasses raise when provided a dtype instead of just returning an identical index without doing anything?

**Comment by flexatone** [comment]
The `astype` methods in `Index` and `IndexHierarchy` are now more powerful and less surprising! This will be included in the 0.6.11 release.

```python
>>> idx = sf.IndexYear(('2020', '2021'))
>>> idx
<IndexYear>
2020
2021
<datetime64[Y]>
>>> idx.astype('datetime64[D]')
<IndexDate>
2020-01-01
2021-01-01
<datetime64[D]>
>>> 
>>> idx = sf.IndexHierarchy.from_product(('2020','2021'), (10, 20))
>>> idx
<IndexHierarchy>
2020             10
2020             20
2021             10
2021             20
<<U4>            <int64>
>>> idx.astype[0]('datetime64[D]')
<IndexHierarchy>
2020-01-01       10
2020-01-01       20
2021-01-01       10
2021-01-01       20
<datetime64[D]>  <int64>
```
